### PR TITLE
Feature/Abi decoder

### PR DESCRIPTION
**Pull request by yehor-podporinov**

### CODE REVIEW

File: forms/AbiDecodeForm.vue [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+  const url = 'https://api.openchain.xyz/signature-database/v1/lookup'
+  const query = `function=${selector}`
+
+  const response = await (await fetch(`${url}?${query}`)).json()
+  if (!response.ok) throw new Error(response.error)
+
+  return response.result.function[selector]?.[0]?.name || ''
+}
+
+const decode = async (): Promise<DecodedData> => {
```
Comment (nikita1tup) [inline forms/AbiDecodeForm.vue:170]: change name

File: forms/AbiDecodeForm.vue [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  isDecoding.value = true
+
+  let types: DecodedData['types'] = []
+  let values: DecodedData['values'] = []
+
+  try {
+    switch (true) {
```
Comment (nikita1tup) [inline forms/AbiDecodeForm.vue:177]: to discuss

File: forms/AbiDecodeForm.vue [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      case form.decodeMode === DECODE_MODES.auto && form.hasFuncSelector: {
+        const funcSelector = form.abiEncoding.substring(0, 10)
+        const funcData = '0x' + form.abiEncoding.substring(10)
+
+        const funcSignature = await fetchFuncSignature(funcSelector)
+
+        const funcFragment = funcSignature
+          ? FunctionFragment.from(funcSignature)
+          : guessFragment(form.abiEncoding)
+        if (!funcFragment) throw new Error('failed guess fragment')
```
Comment (nikita1tup) [inline forms/AbiDecodeForm.vue:187]: explain the error

File: forms/AbiDecodeForm.vue (hunk 2 above)
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:177]: discussed

File: forms/AbiDecodeForm.vue (hunk 3 above)
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:187]: discussed

File: forms/AbiDecodeForm.vue [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    form.args = types.map((type, idx) => ({
+      id: form.args[idx]?.id || uuidv4(),
+      type: type,
+      value: formatValue(values[idx]),
+    }))
+
+    errorMessage.value = ''
+  } catch (error) {
+    if (form.decodeMode === DECODE_MODES.auto) form.args.length = 0
+    else form.args = form.args.map(arg => ({ ...arg, value: '' }))
```
[the remark this answers is not in the document]
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:258]: it's the same

File: forms/AbiDecodeForm.vue [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+const formatValue = (value: ArrayElement<DecodedData['values']>): string => {
+  if (value instanceof Array) {
+    const values = value.map(v => (checkIsBigInt(v) ? v.toString() : v))
+    return JSON.stringify(values)
+  }
+
+  return String(value)
+}
+
+let _formStateJsonString = ''
```
[the remark this answers is not in the document]
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:235]: ref is not needed

File: forms/AbiDecodeForm.vue [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        v-model="form.hasFuncSelector"
+        :label="$t('abi-decode-form.has-func-selector-label')"
+      />
+      <radio-button-field
+        v-model="form.decodeMode"
+        :options="decodeModeOptions"
+      />
+    </div>
+    <div
+      v-if="isDecoding || errorMessage || form.args.length"
```
[the remark this answers is not in the document]
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:22]: discussed

File: fields/RadioButtonField.vue [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    filter: grayscale(50);
+    opacity: 0.5;
+  }
+}
+
+.radio-button-field__input {
+  position: absolute;
+  width: toRem(1);
+  height: toRem(1);
+  margin: calc(#{toRem(1)} * -1);
```
[the remark this answers is not in the document]
Comment (yehor-podporinov) [reply fields/RadioButtonField.vue]: yes, done

File: forms/AbiDecodeForm.vue (hunk 1 above)
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue:170]: ?

File: forms/AbiDecodeForm.vue (hunk 3 above)
Comment (ArtSuslov) [reply forms/AbiDecodeForm.vue:187]: 1. it's not localised
2. I don't know how "guessing fragment" works. Do you process this error somewhere else? If not - this description is not enough.
3. why there is no default statement?

File: plugins/localization/resources/en.json [hunk 8]
```diff
@@ -70,6 +70,21 @@
     "date-form-tab": "Human date UTC",
     "timestamp-form-tab": "Timestamp"
   },
+  "abi-decode-form": {
+    "abi-encoding-label": "Abi encoding",
+    "abi-encoding-placeholder": "Abi encoding",
+    "has-func-selector-label": "Assume first four bytes are function selector",
+    "decode-mode-option-title--auto": "Automatically determine the best parameter types",
+    "decode-mode-option-title--manual": "Manually specify parameter types",
```
Comment (ArtSuslov) [inline plugins/localization/resources/en.json:78]: is it BEM in localisation?:)

**Review by ArtSuslov — CHANGES_REQUESTED** [review]

File: plugins/localization/resources/en.json (hunk 8 above)
Comment (yehor-podporinov) [reply plugins/localization/resources/en.json:78]: discussed

File: forms/AbiDecodeForm.vue (hunk 3 above)
Comment (ArtSuslov) [inline forms/AbiDecodeForm.vue:187]: 1. it's not localised
2. I don't know how "guessing fragment" works. Do you process this error somewhere else? If not - this description is not enough.
3. why there is no default statement?

**Review by ArtSuslov — CHANGES_REQUESTED** [review]

File: forms/AbiDecodeForm.vue (hunk 3 above)
Comment (yehor-podporinov) [reply forms/AbiDecodeForm.vue]: TODO: errors localization

**Pull request merged**